### PR TITLE
NOTIF-373 Introduce EphemeralDataInitializer

### PR DESCRIPTION
**Pull request by gwenneg**
This PR introduces data initialization on ephemeral environments.

▶️ **Which data?**

The init will persist any provided `Bundle`, `Application` or `EventType`.

We can't initialize any other kind of data because it would require an `accountId` which can only be found when a REST API is called from the UI (or the python helpers) or when we receive a Kafka message.

▶️ **When?**

The init will only happen if the `ENV_NAME` environment variable contains a value that starts with `env-ephemeral`.

It is guaranteed to happen _after_ the Flyway DB migration is done. For now, we're migrating the DB with [FlywayWorkaround](https://github.com/RedHatInsights/notifications-backend/blob/master/backend/src/main/java/com/redhat/cloud/notifications/db/FlywayWorkaround.java) which observes `StartupEvent`, so I introduced `FlywayEndEvent` to enforce an order and make sure `FlywayWorkaround` runs first. `FlywayWorkaround` will be removed in the future and this won't break the enforced execution order because the Flyway execution from the `quarkus-flyway` extension happens before the beans observing `StartupEvent` are invoked.

▶️ **How?**

There are two ways to init data on ephemeral:
- Persistent: commit a change to the `src/main/resources/ephemeral/ephemeral_data.json` file.
- Temporary: put the data in the `NOTIFICATIONS_EPHEMERAL_DATA` environment variable.

Both can be used at the same time.

Since the `src/main/resources/ephemeral/ephemeral_data.json` file is currently empty, a data example is provided in  `src/main/resources/ephemeral/ephemeral_data.sample.json`.

Here's an example of data passed to the app with the environment variable:
```
NOTIFICATIONS_EPHEMERAL_DATA="{\"bundles\":[{\"name\":\"my-bundle\",\"display_name\":\"My Bundle\",\"applications\":[{\"name\":\"my-app\",\"display_name\":\"My Application\",\"event_types\":[{\"name\":\"my-event-type\",\"display_name\":\"My Event Type\",\"description\":\"This is my event type\"}]}]}]}"
```

▶️ **Why use JSON for the input data?**

Because it will be easier to use for developers who are not familiar with the programming languages we use in notifications.

### CODE REVIEW

**Comment by gwenneg** [comment]
I have a test ready, but the JVM isn't very friendly with env vars changes in tests (unless we add a new dependency) 😢

**Comment by gwenneg** [comment]
I tried a lot of things/workarounds to test this PR, including the [junit-pioneer](https://search.maven.org/artifact/org.junit-pioneer/junit-pioneer/1.4.2/jar) extension. Nothing worked. That's because of the nature of `EphemeralDataInitializer`: it reads the environment variables at application startup. Changes to these variables applied during tests happen too late.

**Comment by pilhuhn** [comment]
This is only a feeling and not a hard fact (or request for change): I'd prefer that the data comes from a ConfigMap and not an env-variable.
The env-variable could though be populated from a ConfigMap.

**Comment by gwenneg** [comment]
@pilhuhn Storing the data in a ConfigMap wouldn't change anything to the current code, would it? Our app would still read the data from the environment variable.

**Comment by pilhuhn** [comment]
Yes, if we use the `valueFrom:` mechanism 

```
env:
        # Define the environment variable
        - name: SPECIAL_LEVEL_KEY
          valueFrom:
            configMapKeyRef:
              # The ConfigMap containing the value you want to assign to SPECIAL_LEVEL_KEY
              name: special-config
              # Specify the key associated with the value
              key: special.how

```
(from https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-container-environment-variables-using-configmap-data )

**Comment by josejulio** [comment]
> Here's an example of data passed to the app with the environment variable:
> 
> ```
> NOTIFICATIONS_EPHEMERAL_DATA="{\"bundles\":[{\"name\":\"my-bundle\",\"display_name\":\"My Bundle\",\"applications\":[{\"name\":\"my-app\",\"display_name\":\"My Application\",\"event_types\":[{\"name\":\"my-event-type\",\"display_name\":\"My Event Type\",\"description\":\"This is my event type\"}]}]}]}"
> ```

Could you document that into the README. Would be great if you added a config-map example too

**Comment by gwenneg** [comment]
Sure.

**Comment by gwenneg** [comment]
Is this what you had in mind @pilhuhn?

```
[...]
- apiVersion: v1
  kind: ConfigMap
  metadata:
    name: notifications-ephemeral-data
  data:
    ephemeral-data: |
      {
        "bundles": [
          {
            "name": "my-bundle",
            "display_name": "My Bundle",
            "applications": [
              {
                "name": "my-app",
                "display_name": "My Application",
                "event_types": [
                  {
                    "name": "my-event-type",
                    "display_name": "My Event Type",
                    "description": "This is my event type"
                  }
                ]
              }
            ]
          }
        ]
      }
[...]
        env:
        - name: NOTIFICATIONS_EPHEMERAL_DATA
          valueFrom:
            configMapKeyRef:
              name: notifications-ephemeral-data
              key: ephemeral-data
[...]
```
OpenShift is happy with that and the env variable looks OK, but I have yet to test it with this PR.

**Comment by gwenneg** [comment]
Doc updated: [deploying-the-backend-with-data-on-ephemeral](https://github.com/RedHatInsights/notifications-backend/blob/3f2a74ac877da6ee9bacbdb3bcaa2156567b04a2/README.adoc#deploying-the-backend-with-data-on-ephemeral)

File: backend/src/main/java/com/redhat/cloud/notifications/epehemeral/Application.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.redhat.cloud.notifications.epehemeral;
```
Comment (josejulio) [inline backend/src/main/java/com/redhat/cloud/notifications/epehemeral/Application.java:1]: small typo on the package name: should be `ephemeral` and not  ep**e**hemeral

**Review by josejulio — APPROVED** [review]
LGTM, small typo on the package name

**Comment by gwenneg** [comment]
> LGTM, small typo on the package name

Good catch, that's fixed.

**Comment by gwenneg** [comment]
I renamed the `ephemeral-data` ConfigMap key to `ephemeral_data.json`.

**Review by pilhuhn — APPROVED** [review]
Perhaps add one line (to the README) about what happens if env-var and internal file are present

**Pull request merged**